runtimes/rust/lbr-prelude/src/json.rs

```diff
@@ -0,0 +1,327 @@
+            ("Nothing", ctor_fields) => 
+                match &ctor_fields[..] {
+                    [] => Ok(None),
+                    _ => Err(serde::ser::Error::custom(format!(
+                        "Expected a JSON Array with 0 fields but got {:?}",
+                        ctor_fields
+                    ))),
+                },
+            ("Just", ctor_fields) => 
+                match &ctor_fields[..] {
+                    [val] => Ok(Some(T::from_json(val.clone())?)),
+                    _ => Err(serde::ser::Error::custom(format!(
+                        "Expected a JSON Array with 1 fields but got {:?}",
+                        ctor_fields
+                    ))),
+                },
+            _ => error(
+                    "Expected a JSON String to contain one of constructor names (Just, Nothing) but got an unknown constructor name.")
+        }
+    }
+}
+
+impl<T, E> Json for Result<T, E>
+where
+    T: Json,
+    E: Json,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        match self {
+            Ok(val) => Ok(sum_constructor("Right", vec![val.to_json()?])),
+            Err(val) => Ok(sum_constructor("Left", vec![val.to_json()?])),
+        }
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        let map = sum_parser(&value)?;
+
+        match map {
+            ("Right", ctor_fields) => 
+                match &ctor_fields[..] {
+                    [val] => Ok(Ok(T::from_json(val.clone())?)),
+                    _ => error(&format!("Expected a JSON Array with 1 fields but got {:?}", ctor_fields)),
+                },
+                ("Left", ctor_fields) => 
+                    match &ctor_fields[..] {
+                        [val] => Ok(Err(E::from_json(val.clone())?)),
+                        _ =>  error(&format!(
+                                        "Expected a JSON Array with 1 fields but got {:?}",
+                                        ctor_fields
+                                        )),
+                    },
+                _ => error(
+                        "Expected a JSON String to contain one of constructor names (Just, Nothing) but got an unknown constructor name.")
+        }
+    }
+}
+
+impl<T> Json for Vec<T>
+where
+    T: Json,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        let values = self
+            .iter()
+            .map(|val| val.to_json())
+            .into_iter()
+            .collect::<Result<Vec<Value>, Error>>()?;
+
+        Ok(Value::Array(values))
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Array(vec) => vec
+                .iter()
+                .map(|val| T::from_json(val.clone()))
+                .into_iter()
+                .collect::<Result<Vec<T>, Error>>(),
+            _ => error(&format!("Expected a JSON Array but got {:?}", value)),
+        }
+    }
+}
+
+impl<T> Json for HashSet<T>
+where
+    T: Json + Eq + Hash,
```

Comment:
Addition of `Hash` here probably seems benign, but that means that during code generation, we have to derive Hash for all LB types we map into Rust.

Feel free to add a `NOTE(syg251): Must add Hash derivation to type def printing`

Reply:
> Fyi, I'm using this notation for comments that can be collected and analyzed by 3rd party tooling

That's nice, I wanted to introduce something like that in another project

If we use HashSet (which seems to be a reasonable choice), we must have the Hash trait implemented for the key. But now that you mention it, I think we could provide a more generic implementation for anything implementing the Iterator trait. This would tick out Vec, HashSet and HashMap, and a lot more. I'll play around with this for a bit...

Reply:
I'm thinking about using BTreeMap instead of a HashSet as a default set implementation, but I think it's not the common choice.
Using BTreeMap would have these two benefits:
- same constraints apply as in HS and PS (Ord instead of Hash)
- json instanced are "automatically" come out ordered
What do you think?

Reply:
cp from the chat

You decide, it's ok to do whatever and adjust after if needed. I think Prelude.Map encoding needs to be documented. I 'think' that by accident the keys are ordered, and I would say "let's make that part of the LB.Json encoding standard for Map".
So {b: 1, a: 2} dictionary is serialized as [["a", 2], ["b", 1]].